tracing-tracy/src/lib.rs

```diff
@@ -228,55 +257,86 @@ where
                     self.stack_depth,
                 ),
                 id.into_u64(),
-            ));
+            )
+        };
+
+        TRACY_SPAN_STACK.with(|s| {
+            s.push(stack_frame);
         });
     }
 
     fn on_exit(&self, id: &Id, _: Context<S>) {
-        TRACY_SPAN_STACK.with(|s| {
-            if let Some((span, span_id)) = s.borrow_mut().pop_back() {
-                if id.into_u64() != span_id {
-                    self.client.color_message(
-                        "Tracing spans exited out of order! \
-                        Trace may not be accurate for this span stack.",
-                        0xFF000000,
-                        self.stack_depth,
-                    );
-                }
-                drop(span);
-            } else {
+        let stack_frame = TRACY_SPAN_STACK.with(|s| s.pop());
+
+        if let Some((span, span_id)) = stack_frame {
+            if id.into_u64() != span_id {
                 self.client.color_message(
-                    "Exiting a tracing span, but got nothing on the tracy span stack!",
+                    "Tracing spans exited out of order! \
+                        Trace may not be accurate for this span stack.",
                     0xFF000000,
                     self.stack_depth,
                 );
             }
-        });
+            drop(span);
+        } else {
+            self.client.color_message(
+                "Exiting a tracing span, but got nothing on the tracy span stack!",
+                0xFF000000,
+                self.stack_depth,
+            );
+        }
+    }
+
+    fn on_close(&self, id: Id, ctx: Context<'_, S>) {
+        let Some(span) = ctx.span(&id) else { return };
+
+        if let Some(fields) = span.extensions_mut().get_mut::<FormattedFields<F>>() {
+            let buf = mem::take(&mut fields.fields);
+            CACHE.with(|cache| drop(StrCacheGuard::new(cache, buf)));
+        };
     }
 }
 
-struct TracyEventFieldVisitor {
-    dest: String,
+struct TracyEventFieldVisitor<'a> {
+    dest: &'a mut String,
     frame_mark: bool,
     first: bool,
 }
 
-impl Visit for TracyEventFieldVisitor {
+impl Visit for TracyEventFieldVisitor<'_> {
     fn record_bool(&mut self, field: &Field, value: bool) {
         match (value, field.name()) {
-            (true, "tracy.frame_mark") => self.frame_mark = true,
-            _ => self.record_debug(field, &value),
+            (_, "tracy.frame_mark") => self.frame_mark = value,
```

Comment:
I changed this to set the value... was the previous behavior intentional?

Reply:
In principle a field name for an event can occur only once. I don’t particularly care either way, but `tracy.frame_mark = false` does not really have any reason to show up in the user code… so it seemed to make sense to not invoke any special logic on it? Unless I guess they want to make it dynamic somehow? :shrug:

Reply:
> Unless I guess they want to make it dynamic somehow?

Yeah that's all I could think of. It just felt wrong to me to take in a bool and ignore it. I could see it as being confusing behavior to track down if you're expecting false to work for whatever reason.